src/vs/workbench/parts/search/browser/openRepoHandler.ts

```diff
@@ -0,0 +1,301 @@
+	}
+
+	public onOpen(): void {
+		this.cacheState = new CacheState(cacheKey => this.cacheQuery(cacheKey), query => this.searchService.searchRepo(query), cacheKey => this.searchService.clearCache(cacheKey), this.cacheState);
+		this.cacheState.load();
+	}
+
+	private cacheQuery(cacheKey: string): ISearchQuery {
+		const options: IQueryOptions = {
+			folderResources: this.contextService.getWorkspace() ? [this.contextService.getWorkspace().resource] : [],
+			extraFileResources: getOutOfWorkspaceEditorResources(this.editorGroupService, this.contextService),
+			filePattern: '',
+			cacheKey: cacheKey,
+			maxResults: 0,
+			sortByScore: true
+		};
+
+		const query = this.queryBuilder.file(options);
+		this.searchService.extendQuery(query);
+
+		return query;
+	}
+
+	public get isCacheLoaded(): boolean {
+		return this.cacheState && this.cacheState.isLoaded;
+	}
+
+	public getGroupLabel(): string {
+		return nls.localize('searchResults', "search results");
+	}
+
+	public getAutoFocus(searchValue: string): IAutoFocus {
+		return {
+			autoFocusFirstEntry: true
+		};
+	}
+}
+
+enum LoadingPhase {
+	Created = 1,
+	Loading,
+	Loaded,
+	Errored,
+	Disposed
+}
+
+/**
+ * Exported for testing.
+ */
+export class CacheState {
+
+	private _cacheKey = uuid.generateUuid();
+	private query: ISearchQuery;
+
+	private loadingPhase = LoadingPhase.Created;
+	private promise: TPromise<void>;
+
+	constructor(cacheQuery: (cacheKey: string) => ISearchQuery, private doLoad: (query: ISearchQuery) => TPromise<any>, private doDispose: (cacheKey: string) => TPromise<void>, private previous: CacheState) {
+		this.query = cacheQuery(this._cacheKey);
+		if (this.previous) {
+			const current = objects.assign({}, this.query, { cacheKey: null });
+			const previous = objects.assign({}, this.previous.query, { cacheKey: null });
+			if (!objects.equals(current, previous)) {
+				this.previous.dispose();
+				this.previous = null;
+			}
+		}
+	}
+
+	public get cacheKey(): string {
+		return this.loadingPhase === LoadingPhase.Loaded || !this.previous ? this._cacheKey : this.previous.cacheKey;
+	}
+
+	public get isLoaded(): boolean {
+		const isLoaded = this.loadingPhase === LoadingPhase.Loaded;
+		return isLoaded || !this.previous ? isLoaded : this.previous.isLoaded;
+	}
+
+	public get isUpdating(): boolean {
+		const isUpdating = this.loadingPhase === LoadingPhase.Loading;
+		return isUpdating || !this.previous ? isUpdating : this.previous.isUpdating;
+	}
+
+	public load(): void {
+		if (this.isUpdating) {
+			return;
+		}
+		this.loadingPhase = LoadingPhase.Loading;
+		this.promise = this.doLoad(this.query)
+			.then(() => {
+				this.loadingPhase = LoadingPhase.Loaded;
+				if (this.previous) {
+					this.previous.dispose();
+					this.previous = null;
+				}
+			}, err => {
+				this.loadingPhase = LoadingPhase.Errored;
+				errors.onUnexpectedError(err);
+			});
+	}
+
+	public dispose(): void {
+		if (this.promise) {
+			this.promise.then(null, () => { })
```

Comment:
It may suppress a "promise not handled" warning from vscode's promise system, which may or may not be a good thing